text/0026-transform_application.md

```diff
@@ -0,0 +1,107 @@
+
+Getting down to it, there's a fundamental mismatch with how transforms are defined, and how FSI assumes it may operate. Before FSI, there was a one-to-one correlation between running a transform and creating a commit, hence why it made sense to tie that execution to the "save" command. The `--dry-run` flag was acting as a development tool. But with FSI, that original assumption no longer applies. A transform should be able to be executed as a way to modify the body without committing, in fact this is probably going to be the expected behavior for any new user coming to qri.
+
+# Proposal
+
+Transforms should not be thought of as equivalent to a state transition to a new commit. They should be viewed literally as a *transform*, from some input to some output, and only sometimes is that performed as creating a commit.
+
+Let's say we add a command `qri apply`, which runs a transform to arrive at some output; but how that output is applies depends upon the context of the command.
+
+1) When operated in the context of a manual save, `apply` can write to stdout, replacing the need for `--dry-run`. In this case, we know that no commit is written, so we can change `save` to write to stdout in other cases.
+2) A transform as a state transition can be thought of as a `qri apply` operation piped into the `qri save` command, with appropriate syntacic sugar.
+3) For FSI, `qri apply` can modify the body in the working directory, letting the user see the results without messing with the `save` and `--dry-run` comamnds. When they eventually do run `qri save`, we should stash the current body file, apply the transform and verify one of the following:
+  * The stashed body matches the output of the transform
+  * The stashed body is unchanged from the previous commit, replace it with the transforms' output
+  * Otherwise throw an error because the user may be making a mistake
+4) Not mentioned above, but once Desktop has better support for transform, we probably will eventually want some interactive development environment. Perhaps transforms can be incrementally compiled and applied, resulting in fast turnaround.
+
+To expand on this idea, applying a transform could be generalized out to other tasks that execute some sort of data pipeline. For example, if we bring back viz, generating those visuals could also be seen as a type of apply operation, with the script being some sort of imaging library, and the output being static images. Or, our `qri update` feature could be somehow incorporated into this. Any sort of task that computes using a dataset to produce some view or artifact could be folded into this command. Better to have a generalized tool than create a one-off command that solves only a narrow problem.
```

Comment:
_Ugh this comment is taking too long to write! I'll try to come back & finish this_

I really like this, and feel like you've hit on a big concept here @dustmop. I think of `apply` as `execute scripts on datasets`. One could also say "apply transforms datasets". We can even get formal about it conceiving of as apply as executing a single function with one or more datasets as input, and returns a single function.

```
apply = fn(ds1, ds2, ...): ds 
```

Now, that function can _absolutely_ have side effects (making network calls is a side effect), Capturing as many of those side effects as possible in the resulting dataset

I totally think we should make the area `apply` covers large, extending it to all "executable input". Apply may be a 

We write lots of functions within the qri codebase itself that would qualify as "apply"

```
sql
diff
render
stats
validate
```

This is mainly so we can think about it formally, where apply

I'm not saying any of those

Reply:
I'd like to double down on this. Actually for the longer term I think it would be good to have all "modifier" operations be part of the apply ecosystem, governed by the "applicator engine" of qri. Ideally this would detach data storage/versioning from any other operations which would split the responsibility of qri on data management and data processing into distinct buckets. Hopefully this would allow for writing very custom toolchains later on top of the "apply" paradigm and potentially be exposed to the user as well to suit their custom needs.

Reply:
I think we need to constrain the definition of `apply` somewhat, so that it doesn't become simply synonymous with "code", and therefore lose all useful meaning. To do that, we need to figure out what's the benefit of defining `application`, what sort of guarantees do we want from it, as something that acts like running a transform, but is more general.

Suggestions: 
- `apply` uses a dataset as input, possibly has other inputs, and the output is something that can become a component of a dataset. So something like performance measurements wouldn't qualify. I don't think `validate` does either, but am not sure.
- `apply` is always optional, but is a separate code path from a normal save. This means `diff` doesn't qualify, because we actually use `diff` when making a commit, both to see if anything changed, and to calculate the friendly commit message.
- The others, `sql`, `render`, `stats` are all good by me.

Reply:
Hmm, I agree. Cramming everything into apply now just gets us back to the initial problem just at a different subcomand. 
I'm all for it to carve out a clear definition of apply and what can fall under it.

If I had to pick, your first suggestion of apply makes the most sense to me. And `validate`, unless the output is saved should not be an `apply` operator anyways, it's result is transient for the most part and think `diff` should also fall into this bracket. It offers valid and useful info but is not part of the dataset, it just describes the transition similar as to how `validate` describes the current state of the dataset.

Reply:
I'm in on having `validate` and `diff` _not_ be part of the apply family. Points made above make a great deal of sense. 

To that end, maybe we drop `stats` as well? Yes stats could theoretically be the body of a dataset, but that feels a little shoehorned to me.

Feels like a nice grouping to have `render`, `sql`, and starlark transformations in one spot.